### ISSUE TITLE
JdkClientHttpRequestFactory not compatible with BufferingClientHttpRequestFactory -  6.1 M3

### ISSUE THREAD

**Issue by alexcrownus** [spring-projects/spring-framework#31027]
In Spring Framework 6.1 M3, using `BufferingClientHttpRequestFactory` with `JdkClientHttpRequestFactory` throws an exception

`java.lang.IllegalArgumentException: non-positive contentLength: 0`.

The issue seems to be from class `BufferingClientHttpResponseWrapper`, method

```java
    @Override
    public InputStream getBody() throws IOException {
        if (this.body == null) {
            this.body = StreamUtils.copyToByteArray(this.response.getBody()) 
        }
        return new ByteArrayInputStream(this.body);
    }
```

Since `BufferingClientHttpResponseWrapper` is not `public`, I copied the class and was able to fix the issue with

```java
    @Override
    public InputStream getBody() throws IOException {
        if (this.body == null) {
            this.body = StreamUtils.copyToByteArray((this.response.getBody() != null) ? this.response.getBody() : new ByteArrayInputStream(new byte[0]));
        }
        return new ByteArrayInputStream(this.body);
    }
```

Not sure if this is a good fix but it works.

**Comment by sbrannen** [comment]
The contract of `org.springframework.http.HttpInputMessage.getBody()` states that it can never return `null`.

So, perhaps there is a concurrency issue with the `body` field in `BufferingClientHttpResponseWrapper` (with regard to the implementation of `BufferingClientHttpResponseWrapper#getBody()`).

Can you provide a minimal example that demonstrates the behavior you've described?

Also, a complete stack trace would be helpful.

**Comment by poutsma** [comment]
Indeed, like @sbrannen said, `HttpInputMesage::getBody` should not return `null`. And looking at the `JdkClientHttpResponse` class, you can see that we specifically guard against a `null` body stream: https://github.com/spring-projects/spring-framework/blob/cf75a0901132adf2c65b8ea1f7ead2b0d80875cf/spring-web/src/main/java/org/springframework/http/client/JdkClientHttpResponse.java#L55
